sdk/core/core/internal/az_precondition_internal.h

```diff
@@ -79,20 +79,50 @@ AZ_NODISCARD AZ_INLINE bool az_span_is_valid(az_span span, int32_t min_size, boo
         - If null_is_valid is true and the pointer in the span is null, the size must also be 0.
         - In the case of the pointer not being NULL, the size is greater than or equal to zero.
   */
+
+  uint8_t* const default_init_ptr = az_span_ptr((az_span){ 0 });
   if (null_is_valid)
   {
-    result = ptr == NULL ? span_size == 0 : span_size >= 0;
+    result = (ptr == NULL || ptr == default_init_ptr) ? span_size == 0 : span_size >= 0;
   }
   else
   {
-    result = ptr != NULL && span_size >= 0;
+    result = (ptr != NULL && ptr != default_init_ptr) && span_size >= 0;
+  }
+
+  if (!result)
+  {
+    return false;
+  }
+
+  // Can't wrap over the end of the address space.
+  // The biggest theoretical pointer value is "(void*)~0" (0xFFFF...), which is the end of address
+  // space. We don't attempt to read/write beyond the end of the address space - it is unlikely a
+  // desired behavior, and it is not defined. So, if the span size is greater than the addresses
+  // left until the theoretical end of the address space, it is not a valid span.
+  // Example: (az_span) { .ptr = (uint8_t*)(~0 - 5), .size = 10 } is not a valid span, because most
+  // likely you end up pointing to 0x0000 at .ptr[6], &.ptr[7] is 0x...0001, etc.
+  uint8_t* const max_ptr = (uint8_t*)~0;
+  if ((size_t)span_size > (size_t)(max_ptr - ptr))
+  {
+    return false;
   }
 
-  return result && min_size <= span_size;
+  return min_size <= span_size;
 }
 
 #define _az_PRECONDITION_VALID_SPAN(span, min_size, null_is_valid) \
-  _az_PRECONDITION(az_span_is_valid(span, min_size, null_is_valid))
+  _az_PRECONDITION(_az_span_is_valid(span, min_size, null_is_valid))
+
+AZ_NODISCARD AZ_INLINE bool _az_span_overlap(az_span a, az_span b)
+{
+  uint8_t* const a_ptr = az_span_ptr(a);
+  uint8_t* const b_ptr = az_span_ptr(b);
+
+  return a_ptr <= b_ptr ? (a_ptr + az_span_size(a) > b_ptr) : (b_ptr + az_span_size(b) > a_ptr);
```

Comment:
Does an empty span overlap with any span (empty or non empty)? I say no.

Add checks:
```C
if (az_span_size(a) == 0 || az_span_size(b) == 0)
  return false;
```

Also, does it make sense to consider `a_ptr + az_span_size(a) == b_ptr` to be overlapping?

Reply:
Also, is there a way to support overlapping spans when doing url encode?

One concern with this check is that now the following scenario isn't allowed, even though it is "safe":
```C
buffer[100];
buffer[99] = 'A';
dest = buffer;
src = buffer.Slice(99, 100);

// fails because the spans overlap, but the dest has plenty of space to hold the src.
url_encode(dest, src);
```

Maybe that's OK, just asking to see how feasible it would be.

Reply:
There might be some cases when source and destination may overlap, and everything is going to be just fine. It is harder and IMO not with the cost to check for all these cases.

When it is safe to encode in the same buffer, it should be possible to write a code that clearly slices source and destination into two non-overlapping buffers, and calls the function. This way, you are always guaranteed to not overrun the buffer and get hardly predictable results, when the nature of the input changes.

Other case would be to pass two spans pointing to the same buffer with the contents of "abc". Today it is fine, tomorrow you change "abc" to "a/c", and may end up with unexpected behavior. Preconditions will help you to catch it earlier and either reword your intent more explicitly, or avoid the unexpected behavior in all cases.

`a_ptr + az_span_size(a) == b_ptr` are not overlapping because buffer `b` starts just as the buffer `a` ends, they are right next to each other, but don't overlap.

As of the empty buffer pointing inside another buffer being considered overlapping, that was a little dilemma, but I decided to keep it this way for the same reason we don't allow NULL destination spans even when the source is empty. Should you change the size of the inner span from 0 to 1, they will start to overlap. It is a slight difference, and we can remove that restriction once we have a data to back up that we really don't want the buffer to be considered overlapping in this case.
Interestingly, two empty buffers with the same ptr, or an empty span just at the end of the non-empty span are not considered overlapping, but being next to each other.